File: sdk/jvm/src/test/java/com/ttzip/InteropCli.java
// SPDX-License-Identifier: BSD-3-Clause OR Apache-2.0
//
// Copyright (c) 2026 Witt Kung <witt.w.kung@gmail.com> and TTZip Contributors.
// All rights reserved.
//
// TTZip: High-performance native archiving and compression engine for Java 22+.
// Headless Panama FFM Interop CLI Runner.

package com.ttzip;

import java.util.List;

public final class InteropCli {

    private static TTZip.ArchiveFormat parseFormat(String fmtStr) {
        if (fmtStr == null) return TTZip.ArchiveFormat.ZIP;
        String s = fmtStr.toLowerCase();
        return switch (s) {
            case "zip" -> TTZip.ArchiveFormat.ZIP;
            case "7z", "7zip", "sevenzip" -> TTZip.ArchiveFormat.SEVEN_ZIP;
            case "tar" -> TTZip.ArchiveFormat.TAR;
            case "tar.gz", "targz", "tgz", "gz" -> TTZip.ArchiveFormat.TAR_GZ;
            case "tar.bz2", "tarbz2", "tbz2", "bz2" -> TTZip.ArchiveFormat.TAR_BZ2;
            case "tar.xz", "tarxz", "txz", "xz" -> TTZip.ArchiveFormat.TAR_XZ;
            case "tar.zst", "tarzst", "tar.zstd", "zst" -> TTZip.ArchiveFormat.TAR_ZSTD;
            default -> TTZip.ArchiveFormat.ZIP;
        };
    }

    private static void printUsage() {
        System.err.println("Usage:");
        System.err.println("  InteropCli --create <format> <src> <dst> [--password <pwd>]");
        System.err.println("  InteropCli --extract <src> <dst> [--password <pwd>]");
        System.err.println("  InteropCli --version");
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            printUsage();
            System.exit(2);
        }

        if (args[0].equals("--version")) {
            System.out.println(TTZip.version());
            System.exit(0);
        }

        String mode = null;
        String formatStr = null;
        String src = null;
        String dst = null;
        String password = null;

        int i = 0;
        while (i < args.length) {
            String arg = args[i];
            switch (arg) {
                case "--create" -> {
                    mode = "create";
                    if (i + 3 >= args.length) {
                        System.err.println("Error: --create requires <format> <src> <dst>");
                        System.exit(2);
                    }
                    formatStr = args[i + 1];
                    src = args[i + 2];
                    dst = args[i + 3];
                    i += 4;
                }
                case "--extract" -> {
                    mode = "extract";
                    if (i + 2 >= args.length) {
                        System.err.println("Error: --extract requires <src> <dst>");
                        System.exit(2);
                    }
                    src = args[i + 1];
                    dst = args[i + 2];
                    i += 3;
                }
                case "--password" -> {
                    if (i + 1 >= args.length) {
                        System.err.println("Error: --password requires an argument");
                        System.exit(2);
                    }
                    password = args[i + 1];
                    i += 2;
                }
                default -> {
                    System.err.println("Unknown argument: " + arg);
                    printUsage();
                    System.exit(2);
                }
            }
        }

        if (mode == null) {
            printUsage();
            System.exit(2);
        }

        try {
            if ("create".equals(mode)) {
                TTZip.ArchiveFormat fmt = parseFormat(formatStr);
                TTZip.compress(
                    List.of(src),
                    dst,
                    fmt,
                    TTZip.CompressionLevel.NORMAL,
                    password,
                    0,
                    null
                );
                System.exit(0);
            } else if ("extract".equals(mode)) {
                TTZip.extract(src, dst, password, 0, null);
                System.exit(0);
            }
        } catch (Throwable t) {
            System.err.println("Error: " + t.getMessage());
            t.printStackTrace(System.err);
            System.exit(1);
        }

        System.exit(2);
    }
}

File: python/interop_cli.py
#!/usr/bin/env python3
# SPDX-License-Identifier: BSD-3-Clause OR Apache-2.0
#
# Copyright (c) 2026 Witt Kung <witt.w.kung@gmail.com>
# All rights reserved.
#
# TTZip: High-performance native archiving and compression engine.

# Python PyO3 Headless Interop CLI Runner.

import sys
import os
from pathlib import Path

# Add python directory to sys.path
SCRIPT_DIR = Path(__file__).resolve().parent
if str(SCRIPT_DIR) not in sys.path:
    sys.path.insert(0, str(SCRIPT_DIR))

import ttzip


def parse_format(fmt_str: str) -> str:
    s = fmt_str.lower()
    if s in ("zip", "pkzip"):
        return "zip"
    if s in ("7z", "7zip", "sevenzip"):
        return "7z"
    if s in ("tar",):
        return "tar"
    if s in ("tar.gz", "targz", "tgz", "gz"):
        return "tar.gz"
    if s in ("tar.bz2", "tarbz2", "tbz2", "bz2"):
        return "tar.bz2"
    if s in ("tar.xz", "tarxz", "txz", "xz"):
        return "tar.xz"
    if s in ("tar.zst", "tarzst", "tar.zstd", "zst"):
        return "tar.zst"
    return "zip"


def print_usage(prog: str) -> None:
    sys.stderr.write(f"Usage:\n")
    sys.stderr.write(f"  {prog} --create <format> <src> <dst> [--password <pwd>]\n")
    sys.stderr.write(f"  {prog} --extract <src> <dst> [--password <pwd>]\n")
    sys.stderr.write(f"  {prog} --version\n")


def main() -> int:
    args = sys.argv[1:]
    if not args:
        print_usage(sys.argv[0])
        return 2

    if args[0] == "--version":
        print(ttzip.version())
        return 0

    mode = None
    format_str = None
    src = None
    dst = None
    password = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == "--create":
            mode = "create"
            if i + 3 >= len(args):
                sys.stderr.write("Error: --create requires <format> <src> <dst>\n")
                return 2
            format_str = args[i + 1]
            src = args[i + 2]
            dst = args[i + 3]
            i += 4
        elif arg == "--extract":
            mode = "extract"
            if i + 2 >= len(args):
                sys.stderr.write("Error: --extract requires <src> <dst>\n")
                return 2
            src = args[i + 1]
            dst = args[i + 2]
            i += 3
        elif arg == "--password":
            if i + 1 >= len(args):
                sys.stderr.write("Error: --password requires an argument\n")
                return 2
            password = args[i + 1]
            i += 2
        else:
            sys.stderr.write(f"Unknown argument: {arg}\n")
            print_usage(sys.argv[0])
            return 2

    if not mode:
        print_usage(sys.argv[0])
        return 2

    try:
        if mode == "create":
            fmt = parse_format(format_str)
            ttzip.compress(
                sources=src,
                destination=dst,
                format=fmt,
                level=6,
                password=password,
            )
            return 0
        elif mode == "extract":
            ttzip.extract(
                archive=src,
                destination=dst,
                password=password,
            )
            return 0
    except Exception as ex:
        sys.stderr.write(f"Error: {ex}\n")
        return 1

    return 2


if __name__ == "__main__":
    sys.exit(main())
